### PR TITLE
Column batch tests

### PR DESCRIPTION
**Pull request by gdynusa**
It's WIP, but please check it now, so if I make any fundamental mistakes, I can avoid making them when continuing with the PR

**Merge with (into) #10,  DO NOT SQUASH!**

### CODE REVIEW

**Comment by blazekadam** [comment]
image

(except for the failed tests and missing newlines and maybe a few more things :) )

File: test/core/stream/column_t.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+BOOST_AUTO_TEST_CASE(test_extract_move_only_column)
+{
+    using hipipe::stream::abstract_column;
+
+    std::unique_ptr<Unique> col = std::make_unique<Unique>();
+    col->data().push_back(std::make_unique<int>(1));
+    col->data().push_back(std::make_unique<int>(2));
+    col->data().push_back(std::make_unique<int>(3));
+    BOOST_TEST(*(col->extract<Unique>().at(0)) == 1);
```
Comment (FloopCZ) [inline test/core/stream/column_t.cpp]: `*` has a lower priority, the extra parentheses are not necessary
```suggestion
    BOOST_TEST(*col->extract<Unique>().at(0) == 1);
```

File: test/core/stream/column_t.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+BOOST_AUTO_TEST_CASE(test_take_column)
+{
+    using hipipe::stream::abstract_column;
+    
+    std::unique_ptr<Int> col1 = std::make_unique<Int>();
+    col1->data().assign({1, 2, 3, 4, 5});
+    std::unique_ptr<abstract_column> col2 = col1->take(2);
+    BOOST_TEST(col1->extract<Int>() == std::vector<int>({3, 4, 5}));
+    BOOST_TEST(col2->extract<Int>() == std::vector<int>({1, 2}));
```
Comment (FloopCZ) [inline test/core/stream/column_t.cpp:82]: I like to test corner cases, can you also test taking all the elements, please? :-)

File: test/core/stream/column_t.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+
+BOOST_AUTO_TEST_CASE(test_take_throws_error)
+{
+    using hipipe::stream::abstract_column;
+
+    std::unique_ptr<Int> col1 = std::make_unique<Int>();
+    col1->data().assign({1, 2, 3, 4, 5});
+    BOOST_CHECK_THROW(col1->take(10), std::runtime_error);
```
Comment (FloopCZ) [inline test/core/stream/column_t.cpp:114]: I like to test corner cases, can you also test that taking 6 throws, please?

File: test/core/stream/column_t.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    col1->data().push_back(std::make_unique<int>(2));
+    std::unique_ptr<Unique> col2 = std::make_unique<Unique>();
+    col2->data().push_back(std::make_unique<int>(3));
+    col1->push_back(std::move(col2));
+
+    BOOST_TEST(col1->size() == 3);
+    BOOST_TEST(*(col1->extract<Unique>().at(0)) == 1);
+    BOOST_TEST(*(col1->extract<Unique>().at(1)) == 2);
+    BOOST_TEST(*(col1->extract<Unique>().at(2)) == 3);
+}
```
Comment (FloopCZ) [inline test/core/stream/column_t.cpp]: Missing newline at the end of file.

File: test/core/stream/column_t.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    col3->data().assign({6, 7});
+
+    col1->push_back(std::move(col2));
+    BOOST_TEST(col1->extract<Int>() == std::vector<int>({1, 2, 3, 4, 5}));
+    col1->push_back(std::move(col3));
+    BOOST_TEST(col1->extract<Int>() == std::vector<int>({1, 2, 3, 4, 5, 6, 7}));
+}
+
+
+BOOST_AUTO_TEST_CASE(test_push_back_only_column)
```
Comment (FloopCZ) [inline test/core/stream/column_t.cpp]:
```suggestion
BOOST_AUTO_TEST_CASE(test_push_back_move_only_column)
```

**Review by FloopCZ — COMMENTED** [review]
@gdynusa I am sincerely impressed. This is exactly as it should be and I would not do it better myself (probably not even this good). And big :+1: for fixing my typos, copy&pastes and even invalid pointer usage in docstring.

File: test/core/stream/batch_t.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+BOOST_AUTO_TEST_CASE(test_batch_insert_assign_extract_move_only)
+{
+    using hipipe::stream::batch_t;
+
+    batch_t batch;
+    batch.insert_or_assign<Unique>();
+    batch.extract<Unique>().push_back(std::make_unique<int>(1));
+    batch.extract<Unique>().push_back(std::make_unique<int>(2));
+    BOOST_TEST(*batch.extract<Unique>().at(0) == 1);
+    BOOST_TEST(*batch.extract<Unique>().at(1) == 2);
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp:54]: When checking by indices, please add 
```c++
BOOST_TEST(batch.extract<Unique>().size() == 2);
```

File: test/core/stream/batch_t.cpp [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    batch.insert_or_assign<Unique>();
+    BOOST_CHECK_THROW(*batch.extract<Unique>().at(0), std::out_of_range);
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp]: Replace by:
```c++
BOOST_TEST(batch.extract<Unique>().empty());
```

File: test/core/stream/batch_t.cpp [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    batch_t batch2;
+    batch2.insert_or_assign<Int>(std::vector<int>{3, 4});
+    batch2.insert_or_assign<Double>(std::vector<double>{0., 1., 2.});
+    batch_t batch3;
+    batch1.push_back(std::move(batch2));
+    BOOST_TEST(batch1.extract<Int>() == std::vector<int>({0, 1, 2, 3, 4}));
+    BOOST_TEST(batch1.extract<Double>() == std::vector<double>({0., 1., 2.}));
+    batch1.push_back(std::move(batch3));
+    BOOST_TEST(batch1.extract<Int>() == std::vector<int>({0, 1, 2, 3, 4}));
+    BOOST_TEST(batch1.extract<Double>() == std::vector<double>({0., 1., 2.}));
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp:172]: I like pushing the empty batch!

File: test/core/stream/batch_t.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    batch_t batch1;
+    batch1.insert_or_assign<Unique>();
+    batch1.extract<Unique>().push_back(std::make_unique<int>(1));
+    batch1.extract<Unique>().push_back(std::make_unique<int>(2));
+    batch_t batch2;
+    batch2.insert_or_assign<Unique>();
+    batch2.extract<Unique>().push_back(std::make_unique<int>(3));
+    batch1.push_back(std::move(batch2));
+    BOOST_TEST(batch1.batch_size() == 3);
+    BOOST_TEST(batch2.batch_size() == 0);
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp]: Please remove this line. The `batch2` is already "moved" and the documentation stays that using it further causes undefined behavior.

**Review by FloopCZ — COMMENTED** [review]
Thank you for your work @gdynusa, this looks great!

**Review by FloopCZ — APPROVED** [review]

**Comment by FloopCZ** [comment]
Thank you @gdynusa, I consider this to be ready and I will merge this when appropriate.
